### PR TITLE
feat: [CDS-59360]: approvals: prevented log stream to close in case of polling errors

### PR DESCRIPTION
**Pull request by naman-goenka**
## Describe your changes
- when an error like connector is invalid etc occurs during approval polling, the log stream is getting closed, for subsequent polling event, continuous stream of log errors is seen due to this.
- for all approvals, in all cases, log stream in only closed in step
- also added appropriate log level for error messages
- jira Approvals bugs resolved: 
 a. On abort, logs  fixed and db state fixed.
 b. on invalid key, npe fixed.
- snow approvals, bugs resolved:
  a. on invalid key, fast fail by improving condition 
 Case when connectors invalid (log stream shouldn't close)
 
- **Testing** - in various cases, close stream is getting called. For polling errors, log stream not getting closed. Valid errors during invalid key in jira and snow approvals. 

Screenshot 2023-04-24 at 12 48 26 PM
Screenshot 2023-04-26 at 1 19 16 PM

Case when invalid ticket number (proper handling)
Screenshot 2023-04-26 at 10 25 43 AM

Screenshot 2023-04-26 at 12 13 43 PM

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger utAll`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/46831)

### CODE REVIEW

**Comment by naman-goenka** [comment]
trigger messagecheck

File: pipeline-service/service/src/main/java/io/harness/pms/approval/AbstractApprovalCallback.java [hunk 1]
```diff
@@ -109,7 +108,7 @@ protected void handleFatalException(
     String errorMessage = String.format(
         "Fatal error evaluating approval/rejection criteria: %s", ngErrorHelper.getErrorSummary(ex.getMessage()));
     logCallback.saveExecutionLog(
-        LogHelper.color(errorMessage, LogColor.Red), LogLevel.INFO, CommandExecutionStatus.FAILURE);
+        LogHelper.color(errorMessage, LogColor.Red), LogLevel.ERROR, CommandExecutionStatus.FAILURE);
```
Comment (puthrayaharness) [inline pipeline-service/service/src/main/java/io/harness/pms/approval/AbstractApprovalCallback.java:111]: Do not send the finalize status. Let the step take care of closing the log stream

File: pipeline-service/service/src/main/java/io/harness/pms/approval/servicenow/ServiceNowApprovalCallback.java [hunk 2]
```diff
@@ -160,7 +162,7 @@ protected void handleFatalException(
     String errorMessage = String.format("Fatal error evaluating approval/rejection/change window criteria: %s",
         ngErrorHelper.getErrorSummary(ex.getMessage()));
     logCallback.saveExecutionLog(
-        LogHelper.color(errorMessage, LogColor.Red), LogLevel.INFO, CommandExecutionStatus.FAILURE);
+        LogHelper.color(errorMessage, LogColor.Red), LogLevel.ERROR, CommandExecutionStatus.FAILURE);
```
Comment (puthrayaharness) [inline pipeline-service/service/src/main/java/io/harness/pms/approval/servicenow/ServiceNowApprovalCallback.java:165]: Do not send finalize status. Let the step take care of closing the stream

**Review by puthrayaharness — CHANGES_REQUESTED** [review]

File: pipeline-service/service/src/main/java/io/harness/pms/approval/AbstractApprovalCallback.java (hunk 1 above)
Comment (naman-goenka) [reply pipeline-service/service/src/main/java/io/harness/pms/approval/AbstractApprovalCallback.java:111]: to confirm in `handleAsyncResponseInternal` method in step, closing will be handled.

File: pipeline-service/service/src/main/java/io/harness/pms/approval/servicenow/ServiceNowApprovalCallback.java (hunk 2 above)
Comment (naman-goenka) [reply pipeline-service/service/src/main/java/io/harness/pms/approval/servicenow/ServiceNowApprovalCallback.java]: done

**Review by puthrayaharness — DISMISSED** [review]

**Review by abhinavmittal-harness — APPROVED** [review]
> Reviewed 8 of 9 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/46831)* status: 9 file(s) to review. 2. unresolved discussions. 1 of 2 approvals obtained.

**Comment by naman-goenka** [comment]
trigger review

**Review by Y0709 — APPROVED** [review]

**Comment by naman-goenka** [comment]
trigger review

**Comment by naman-goenka** [comment]
trigger review

**Review by shivamnegi94 — APPROVED** [review]

**Pull request merged**